### PR TITLE
drivers: flash: spi_nor: Add reset pin

### PR DESCRIPTION
**Pull request by vgeneves**
Add Reset pin initialization during Spi NOR driver start-up. Toggle the pin to reset the device.

### CODE REVIEW

**Review by de-nordic — COMMENTED** [review]
I am not sure whether just existence of reset pin should automatically trigger reset at start.
You may be sharing memory between two MCUs and in this case reset of one MCU may reset flash device in the middle of operation on other MCU.
I am fine with adding the pin and processing but startup reset should be rather conditional.

**Comment by vgeneves** [comment]
Hi,

This implementation is based on the one done on AT45 Spi flash driver (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/spi_flash_at45.c)
Reset pin is optional in the device tree so it can be omitted if its causing problems.

Also, I don't quite understand your use case, spi_nor driver is based on Spi where there can only be one master on the bus so you can't have problems with MCUs accessing the flash at the same time.

Vincent

File: drivers/flash/spi_nor.c [hunk 1]
```diff
@@ -1143,6 +1147,23 @@ static int spi_nor_configure(const struct device *dev)
...
+			return -ENODEV;
+		}
+		if (gpio_pin_configure_dt(cfg->reset, GPIO_OUTPUT_ACTIVE)) {
+			LOG_ERR("Couldn't configure reset pin");
+			return -ENODEV;
+		}
+		rc = gpio_pin_set_dt(cfg->reset, 0);
+		if (rc) {
+			return rc;
+		}
```
Comment (de-nordic) [inline drivers/flash/spi_nor.c:1163]: I guess what happens here is the reset i set active and then inactive immediately after. Shouldn't there be some configurable delay between these two?

**Comment by de-nordic** [comment]
> Hi,
> 
> This implementation is based on the one done on AT45 Spi flash driver (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/spi_flash_at45.c) Reset pin is optional in the device tree so it can be omitted if its causing problems.
> 

I am rather wondering whether the reset should have some kind of property in DTS, where it is defined as "disabled", "on-init" or "on-demand".

> Also, I don't quite understand your use case, spi_nor driver is based on Spi where there can only be one master on the bus so you can't have problems with MCUs accessing the flash at the same time.
>

https://community.infineon.com/t5/Knowledge-Base-Articles/Managing-Two-SPI-Masters-and-One-SPI-Flash-Slave-KBA225990/ta-p/248931

 
> Vincent

**Comment by vgeneves** [comment]
> I am rather wondering whether the reset should have some kind of property in DTS, where it is defined as "disabled", "on-init" or "on-demand".
> 
The reset pin is optional in the device tree, "disabled" and "on-init" are already implemented. "on-demand" seems more complicated and will need a new API.

File: drivers/flash/spi_nor.c (hunk 1 above)
Comment (vgeneves) [reply drivers/flash/spi_nor.c:1163]: On the flash model we are using reset time is 1uS.
But yes it could be configurable in device tree.

**Comment by MaureenHelm** [comment]
@de-nordic any update?

**Review by de-nordic — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**